src/casscf/casscf_gradient.cc

```diff
@@ -0,0 +1,477 @@
+    set_all_variables();
+    write_lagrangian();
+    write_1rdm_spin_dependent();
+    write_2rdm_spin_dependent();
+    tpdm_backtransform();
+    
+    outfile->Printf("\n    Computing Gradient .............................. Done\n");
+    return std::make_shared<Matrix>("nullptr", 0, 0);
+}
+
+
+/**
+ * Write spin_dependent one-RDMs coefficients.
+ *
+ * We force "Da == Db". This function needs be changed if such constraint is revoked.
+ */
+void CASSCF::write_1rdm_spin_dependent() 
+{
+    outfile->Printf("\n    Writing 1RDM Coefficients ....................... ");
+
+    SharedMatrix D1(new Matrix("1rdm coefficients contribution", nirrep_, irrep_vec, irrep_vec));
+
+    for(size_t i = 0, size_c = core_mos_relative.size(); i < size_c; ++i) 
+    {
+        auto m = core_mos_relative[i].second;
+        D1->set(core_mos_relative[i].first, core_mos_relative[i].second, core_mos_relative[i].second, 1.0);
+    }
+
+    (Gamma1_.block("aa")).iterate([&](const std::vector<size_t>& i, double& value)
+    {
+        if (actv_mos_relative[i[0]].first == actv_mos_relative[i[1]].first)
+        {
+            D1->set(actv_mos_relative[i[0]].first, actv_mos_relative[i[0]].second, actv_mos_relative[i[1]].second, value);
+        }
+    });
+
+    D1->back_transform(ints_->Ca());
+    ints_->wfn()->Da()->copy(D1);
+    ints_->wfn()->Db()->copy(D1);
+
+    outfile->Printf("Done");
+}
+
+
+
+/**
+ * Write the Lagrangian.
+ *
+ * This function needs to be changed if frozen approximation is considered!
+ */
+void CASSCF::write_lagrangian() 
+{
+    outfile->Printf("\n    Writing Lagrangian .............................. ");
+
+    set_lagrangian();
+    SharedMatrix L(new Matrix("Lagrangian", nirrep_, irrep_vec, irrep_vec));
+
+    for (const std::string& block : {"cc", "CC", "aa", "AA", "ca", "ac", "CA", "AC"}) 
+    {
+        std::vector<std::vector<std::__1::pair<unsigned long, unsigned long>,
+            std::allocator<std::__1::pair<unsigned long, unsigned long>>>> spin_pair;
+        for (size_t idx: {0, 1})
+        {            
+            auto spin = std::tolower(block.at(idx));
+            if (spin == 'c') { spin_pair.push_back(core_mos_relative);}
+            else if (spin == 'a') { spin_pair.push_back(actv_mos_relative);}
+        }
+
+        (W_.block(block)).iterate([&](const std::vector<size_t>& i, double& value)
+        {           
+            if (spin_pair[0][i[0]].first == spin_pair[1][i[1]].first)
+            {
+                L->add(spin_pair[0][i[0]].first, spin_pair[0][i[0]].second, spin_pair[1][i[1]].second, value);
+            }
+        });                
+    }   
+    L->back_transform(ints_->Ca());
+    ints_->wfn()->Lagrangian()->copy(L);
+
+    outfile->Printf("Done");
+}
+
+
+/**
+ * Write spin_dependent two-RDMs coefficients using IWL.
+ *
+ * Coefficients in d2aa and d2bb need be multiplied with additional 1/2! 
+ */
```

Comment:
Thanks for the implementation note. I do prefer a more detailed version of this about all spin cases and maybe give an example. I think there is a factor of something for d2ab as well.

Reply:
I have added more detailed comments to it.